### PR TITLE
bump-formula|cask-pr: do not allow to bump autobumped packages

### PR DESCRIPTION
**Pull request by iMichka**
Avoids:

- Bumping stuff by mistake that will get autobumped anyway
- Abusing the system to get "free" commits

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (apainintheneck) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:93]:
```suggestion
    if cask.tap.core_cask_tap? && File.readlines(autobump_file_path, chomp: true).include?(cask.name)
```

This way we can avoid the regex entirely.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (apainintheneck) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:116]:
```suggestion
    if formula.tap.core_tap? && File.readlines(autobump_file_path, chomp: true).include?(formula.name)
```

Same reason as above.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (apainintheneck) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:115]: Nit: This could potentially be extracted to `Tap#autobump_file_path`.

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:93]:
```suggestion
    if cask.tap.core_cask_tap? && autobump_file_path.exist? && autobump_file_path).readlines(chomp: true).include?(cask.name)
```

for using `Pathname` methods

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/bump-formula-pr.rb:116]:
```suggestion
    if formula.tap.core_tap? && autobump_file_path.exist? && autobump_file_path).readlines(chomp: true).include?(formula.name)
```

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]:
```suggestion
        Whoops, the #{cask.name} cask has its version update 
        pull requests automatically opened by BrewTestBot!
        We'd still love your contributions, though, so try another one
        that's not in the autobump list:
          #{Formatter.url("https://github.com/Homebrew/homebrew-cask/blob/master/.github/autobump.txt")}
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:119]:
```suggestion
        Whoops, the #{formula.name} formula has its version update 
        pull requests automatically opened by BrewTestBot!
        We'd still love your contributions, though, so try another one
        that's not in the autobump list:
          #{Formatter.url("https://github.com/Homebrew/homebrew-core/blob/master/.github/autobump.txt")}
```

**Review by MikeMcQuaid — COMMENTED** [review]
Great work @iMichka! Some `Pathname` and messaging improvements.

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (ZhongRuoyu) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:95]: > One odd thing (and the reason why used a regex in the first place): `cask.name` returns an array of uppercase cask names.

`cask.token` should be used instead I think:

```suggestion
       autobump_file_path.readlines(chomp: true).include?(cask.token)
```

```
brew(main):001:0> cask = Cask::CaskLoader.load("visual-studio-code")
=> #<Cask visual-studio-code /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask/Casks/v/visual-studio-code.rb>
brew(main):002:0> cask.name
=> ["Microsoft Visual Studio Code", "VS Code"]
brew(main):003:0> cask.token
=> "visual-studio-code"
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (ZhongRuoyu) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:115]:
```suggestion
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (ZhongRuoyu) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:120]:
```suggestion
       autobump_file_path.readlines(chomp: true).include?(formula.name)
```

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (ZhongRuoyu) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:97]:
```suggestion
        Whoops, the #{cask.token} cask has its version update
```

**Review by ZhongRuoyu — COMMENTED** [review]
Seems we also need to allow @BrewTestBot to bypass this restriction.

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:95]: That's it exactly.

**Review by MikeMcQuaid — COMMENTED** [review]
Suggestions look good to me!

**Comment by iMichka** [comment]
This should be good now: thanks for all the suggestions and the help :)

**Review by p-linnane — APPROVED** [review]
Looks great. Thanks @iMichka!

**Review by apainintheneck — APPROVED** [review]

**Review by ZhongRuoyu — COMMENTED** [review]
@BrewTestBot is also using the `bump-formula-pr`/`bump-cask-pr` commands under the hood. Will this prevent it from bumping formulae normally?

**Comment by apainintheneck** [comment]
Looks good to me!

```console
$ brew bump-cask-pr --dry-run --version=4.4.4 tsh
Error: Whoops, the tsh cask has its version update
pull requests automatically opened by BrewTestBot!
We'd still love your contributions, though, so try another one
that's not in the autobump list:
  https://github.com/Homebrew/homebrew-cask/blob/master/.github/autobump.txt
```

```console
$ brew bump-formula-pr --dry-run --version=4.4.4 agda
Error: Whoops, the agda formula has its version update
pull requests automatically opened by BrewTestBot!
We'd still love your contributions, though, so try another one
that's not in the autobump list:
  https://github.com/Homebrew/homebrew-core/blob/master/.github/autobump.txt
```

Thanks @iMichka!

**Comment by apainintheneck** [comment]
> @BrewTestBot is also using this `bump-formula-pr` command under the hood. Will this prevent it from bumping formulae normally?

Good question. I looked at brew test bot and it doesn't seem to use `bump-cask-pr` or `bump-formula-pr` directly outside of the stale PR workflow.

```
/u/l/H/L/T/h/homebrew-test-bot (master|✔) $ git grep bump-cask-pr
.github/workflows/triage-issues.yml:      - name: Mark/Close Stale `bump-formula-pr` and `bump-cask-pr` Pull Requests
.github/workflows/triage-issues.yml:          any-of-labels: "bump-formula-pr,bump-cask-pr"
/u/l/H/L/T/h/homebrew-test-bot (master|✔) $ git grep bump-formula-pr
.github/workflows/triage-issues.yml:      - name: Mark/Close Stale `bump-formula-pr` and `bump-cask-pr` Pull Requests
.github/workflows/triage-issues.yml:          any-of-labels: "bump-formula-pr,bump-cask-pr"
```

I looked at the core repos and it seems we use a separate action to bump packages that calls `brew bump` directly instead of using `brew bump-cask-pr` and `brew bump-formula-pr`.

- https://github.com/Homebrew/actions/tree/master/bump-packages

There is a deprecated action for bumping formulae using `brew bump-formula-pr` but it's not used anymore in core.

- https://github.com/Homebrew/actions/tree/master/bump-formulae

**Comment by ZhongRuoyu** [comment]
`brew bump --open-pr` invokes `brew bump-formula-pr`/`brew bump-cask-pr` under the hood, actually:

https://github.com/Homebrew/brew/blob/154a21706af0e95f0c870b8c358ce483f7065704/Library/Homebrew/dev-cmd/bump.rb#L506-L516

**Comment by apainintheneck** [comment]
Ah, my bad. I had those backwards. I guess we could allow this only on CI somehow.

**Review by apainintheneck — CHANGES_REQUESTED** [review]
We need to find a way to not break CI scripts that run `brew bump`.

**Comment by soft2i** [comment]
[Uploading github-mark-c791e9551fe4.zip…]()

**Comment by soft2i** [comment]
soft2i

**Comment by ZhongRuoyu** [comment]
> We can probably just use `ENV["GITHUB_ACTIONS"]` to check if we're on CI.

Or maybe `HOMEBREW_TEST_BOT_AUTOBUMP`, added in #16664.

**Comment by MikeMcQuaid** [comment]
> Or maybe `HOMEBREW_TEST_BOT_AUTOBUMP`, added in #16664.

This feels like best fit to me 👍🏻

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (ZhongRuoyu) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: Having read #16664 and Homebrew/homebrew-core#163023 more carefully, I think we also need to remove the line below before setting `HOMEBREW_TEST_BOT_AUTOBUMP` in the autobump action:

https://github.com/Homebrew/brew/blob/977ac20b2b491a8df2e28a95541775d0b4aa20fa/Library/Homebrew/utils/github.rb#L573

I'm saying "remove" because it's currently not used and doesn't seem to be useful according to what @Bo98 said in Homebrew/homebrew-core#163023.

(Sorry there've been so many iterations of reviews -- should be my final comment.)

Comment (Bo98) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: Yes

**Comment by apainintheneck** [comment]
I opened a PR to add the environment variable to the `bump-packages` action. After that gets merged in and the cleanup mentioned in https://github.com/Homebrew/brew/pull/16750#discussion_r1508306099 is done, this should be good to go.

- https://github.com/Homebrew/actions/pull/506

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/bump-cask-pr.rb:96]: I opened a PR to rip that out. It looks like there is now no way to skip that check.

- https://github.com/Homebrew/brew/pull/16781

**Comment by apainintheneck** [comment]
Just to over communicate a bit. Three PRs need to be merged in in a specific order and this one will be the last one to land.

See https://github.com/Homebrew/actions/pull/506#issuecomment-1972489707

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:105]:
```suggestion
    if ENV.fetch("HOMEBREW_TEST_BOT_AUTOBUMP", false).blank?
      if cask.tap.core_cask_tap? &&
         (autobump_file_path = cask.tap.path/Tap::HOMEBREW_TAP_AUTOBUMP_FILE) &&
         autobump_file_path.exist? &&
         autobump_file_path.readlines(chomp: true).include?(cask.token)
        odie <<~EOS
          Whoops, the #{cask.token} cask has its version update
          pull requests automatically opened by BrewTestBot!
          We'd still love your contributions, though, so try another one
          that's not in the autobump list:
            #{Formatter.url("#{cask.tap.remote}/blob/master/.github/autobump.txt")}
        EOS
      end
    end
```

let's avoid reading from the file etc. when `HOMEBREW_TEST_BOT_AUTOBUMP` is set

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:128]:
```suggestion
    if ENV.fetch("HOMEBREW_TEST_BOT_AUTOBUMP", nil).blank?
      if formula.tap.core_tap? &&
         (autobump_file_path = formula.tap.path/Tap::HOMEBREW_TAP_AUTOBUMP_FILE) &&
         autobump_file_path.exist? &&
         autobump_file_path.readlines(chomp: true).include?(formula.name)
        odie <<~EOS
          Whoops, the #{formula.name} formula has its version update
          pull requests automatically opened by BrewTestBot!
          We'd still love your contributions, though, so try another one
          that's not in the autobump list:
            #{Formatter.url("#{formula.tap.remote}/blob/master/.github/autobump.txt")}
        EOS
      end
    end
```

File: Library/Homebrew/dev-cmd/bump-cask-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-cask-pr.rb:105]:
```suggestion
    end

```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb:128]:
```suggestion
    end

```

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me now, thanks @iMichka!

**Review by ZhongRuoyu — APPROVED** [review]
Thanks everyone! Let's merge this after Homebrew/homebrew-core#164745 and Homebrew/homebrew-cask#167993.

**Review by p-linnane — APPROVED** [review]
Awesome work everyone!

**Comment by iMichka** [comment]
Thanks all :)

**Pull request merged**
